hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBRWBatchOperation.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package org.apache.hadoop.hdds.utils.db;
+
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.hadoop.hdds.utils.db.RocksDatabase.ColumnFamily;
+import org.apache.hadoop.hdds.utils.db.managed.ManagedReadWriteBatch;
+import org.apache.hadoop.hdds.utils.db.managed.ManagedRocksIterator;
+import org.apache.hadoop.hdds.utils.db.managed.ManagedWriteOptions;
+import org.rocksdb.ColumnFamilyHandle;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.hadoop.hdds.utils.db.managed.ManagedRocksIterator.managed;
+
+/**
+ * Read Write Batch operation implementation for rocks db.
+ */
+public class RDBRWBatchOperation implements RWBatchOperation {
+  public static final Logger LOG =
+      LoggerFactory.getLogger(RDBRWBatchOperation.class);
+  private final ManagedReadWriteBatch writeBatch;
+  
+  private final AtomicLong operationCount = new AtomicLong(0);
+  
+  private final AtomicBoolean isActive = new AtomicBoolean(true);
+
+  private final Object lock = new Object();
+
+  public RDBRWBatchOperation() {
+    writeBatch = new ManagedReadWriteBatch();
+  }
+
+  @Override
+  public void commit(RocksDatabase db) throws IOException {
+    db.batchWrite(writeBatch);
+  }
+
+  @Override
+  public void commit(RocksDatabase db, ManagedWriteOptions writeOptions)
+      throws IOException {
+    db.batchWrite(writeBatch, writeOptions);
+  }
+
+  @Override
+  public void close() {
+    synchronized (lock) {
+      isActive.set(false);
+    }
+
+    waitForNoOperation();
+    writeBatch.close();
+  }
```

Comment:
Instead of spin waiting on close and holding up the caller indefinitely, can we we make close use `operationCount` as a reference counter? Close would either decrement the count if it is non-zero, or close the object if it is zero. The count could be incremented once in the constructor, and once in each call to `newIteratorWithBase`. Then the underlying RocksDB object would be closed on the last user's call to `close`, whether that is the buffer or the iterator. The lock/unlock operation methods would no longer be needed since the class would internalize the functionality.

Reply:
@errose28 Updated with RefCount usages

Reply:
@errose28 Updated with RefCount usages